README.md
**[Requirements](#requirements)** |
**[Usage](#usage)** |
**[Included Tasks](#included-tasks)** |
**[Contribution](#contribution)** |
**[Getting Help](#getting-help)** |
**[TODO](#todo)** |
**[Changelog](#changelog)**

# Cognitive Battery

[![Gitter chat](https://badges.gitter.im/gitterHQ/gitter.png)](https://gitter.im/cognitive-battery/Lobby)

<img align="left" src="images/icon.png">

A Python based battery of common cognitive psychology tasks. Designed to be 
modular as each task is contained within a single Python script/module, the 
results of which are returned as a dataframe for saving.

Pull requests are welcome! Please see the [contribution](#contribution) 
notes below.

<br>

![Battery loading screen](images/battery_screenshot.png)
![Mental rotation task screenshot](images/mrt_screenshot.png)

The battery was originally designed for a resolution of 1280x1024,
but it should work with most resolutions.

Some tasks contain copyrighted images (e.g. Mental Rotations Task, Raven's 
Progressive Matrices) that I cannot include in this repo. In case you have 
access to the original images, I have included an explanation of how to 
name/format those items to work with this battery.

**Note**: I created this project as I needed to get some tasks up and 
running quickly for an experiment. Everything is fully functional as it stands,
but I will be refactoring much of the code over time to clean things up.
This will include things like code reduction, better handling of different
screen resolutions, and improved logic.

## Requirements

Current, and future, versions of the battery use Python 3. Python 2 users can still run v2 of the battery, but will no longer be maintained. Cognitive Battery v3 may still be runnable under Python 2, however, it's not tested or supported, so use at your own risk.

You will need to have the following dependencies installed to run the battery,
most of which are just Python modules:

* **Windows 7+**
  - This battery may work on OSX and Linux, but has only been tested on 
  Windows so far.
* **Python 3.x**
  - The easiest way to install Python is using the 
  [Anaconda](https://www.continuum.io/downloads) distribution as it also 
  includes most of the other dependencies listed below
  - Older versions of the battery (v1-2) are still runnable using Python 2.
* **Pandas**
  - Included with Anaconda. Otherwise, install using pip (`pip install pandas`)
* **Numpy**
  - Included with Anaconda. Otherwise, install using pip (`pip install numpy`)
* **PyQt5**
  - Included with Anaconda. Alternatively, full release (including QT designer)
   downloadable from the
   [PyQT website](https://www.riverbankcomputing.com/software/pyqt/download5)
  - **Note**: Cognitive Battery version 1.x uses PyQt4. However, version 2.x onwards uses PyQt5
* **Pygame**
  - Install using pip (`pip install pygame`)
  - Alternatively, downloadable from the
  [Pygame website](http://www.pygame.org/download.shtml)

## Usage

Using the battery is as simple as running the `run_battery.py` file. This 
can either be done using the command line (navigate to the directory and type
`python run_battery.py`), or running it from IDLE.

In the project manager, add a new bookmark indicating the project name and a path to your desired save directory for that project. Adding or deleting a bookmark does not change any files on the filesystem, it's merely a reference to a directory for where data files and settings should be saved for that project. It's safe to delete a project bookmark as it will not delete any currently existing data.

On the task selection screen, all sections are currently mandatory, although some
 may be irrelevant for your particular experiment. You can place arbitrary 
 values in the irrelevant fields.

Select the tasks you want to include by using the checkboxes. The order of 
task administration can be set using the `Up` and `Down` buttons. 
Alternatively, you can set a random order using the checkbox.

The task results are saved in the `/data` directory of your project directory (specified in the project manager). Each participant's data
 is saved as an Excel file, where each task is saved to a separate sheet.

If you want to reset the settings for a particular project, delete the `battery_settings.ini` file in the project's directory. A new (default) one will be created when you next load that project.

## Included Tasks

Information about the tasks can be found [here](tasks/README.md).

Currently implemented tasks:

* **Attention Network Test (ANT)**
  - Based on the original version by
  [Fan et al. (2002)](http://www.mitpressjournals.org/doi/abs/10.1162/089892902317361886#.VlEwI9irSHs)
* **Digit Span (backwards)**
  - Based on the version used by
  [Berman et al. (2008)](http://pss.sagepub.com/content/19/12/1207) and
  [Bourrier (2015)](https://open.library.ubc.ca/cIRcle/collections/ubctheses/24/items/1.0166677)
* **Eriksen Flanker Task**
  - Based on the compatible and incompatible versions by [Hillman et al. (2006)](http://psycnet.apa.org/record/2006-20659-003) and 
  [Pontifex et al. (2011)](https://www.ncbi.nlm.nih.gov/pubmed/20521857)
* **Mental Rotation Task**
  - Based on the redrawn version by
  [Peters et al. (1995)](http://www.ncbi.nlm.nih.gov/pubmed/7546667)
* **Raven's Progressive Matrices**
  - Based on the second set of matrices used in the
  [Wechsler Adult Intelligence Scale](https://en.wikipedia.org/wiki/Wechsler_Adult_Intelligence_Scale#WAIS-IV)
* **Sternberg Task**
  - Based on the original task by [Sternberg (1966)](https://www.ncbi.nlm.nih.gov/pubmed/5939936),
   with trial count and set sizes from [Martins et al. (2012)](http://www.sciencedirect.com/science/article/pii/S1469029212001380)
* **Sustained Attention to Response Task (SART)**
  - Based on the original version by
  [Robertson et al. (1997)](http://www.sciencedirect.com/science/article/pii/S0028393297000158)

## Contribution

**Note**: I intend to majorly streamline this process in future releases, 
hopefully automating the importing/running/saving of tasks.

Each new task you want to add will need to be programmed as a Python module.

Create a class in the module that contains a `run()` method. This method 
should contain the main sequence of your task and needs to return a Pandas 
dataframe object. Your class should also accept a pygame screen and
background object.

An instance of the class is spawned in `run_battery.py`, near the end of the 
`start()` method. Your class's `run()` method is invoked after, and the 
returned dataframe is saved to file.

You'll then need to update the QT Designer UI file 
(`/designer/ui/battery_window_qt.ui`) by adding your task to the list. 
Rebuild using the included `convertUI.bat` script.

**Note**: It is better to modify the UI file using QT Designer and then 
rebuild, rather than directly editing the generated Python file
(`battery_window_qt.py`), because:

- Your changes will be included in the UI file in case anyone wants to use 
QT Designer to make changes later
- Any changes you make will be lost whenever anyone builds from the UI file

In summary:

1. Program a separate `.py` module for your task
2. Include a `run()` method in your class that contains your task sequence 
and returns a Pandas dataframe
3. Import this module in `run_battery.py`
4. Spawn a new instance of your class in `run_battery.py` in the `start()` 
method
5. Invoke your `run()` method and save the returned dataframe to a file
6. Update the QT Designer UI file and rebuild using the conversion script

Consider making a pull request and please include a journal reference for any
new tasks you add.

## Getting Help
You're encouraged to post questions as an [issue](https://github.com/sho-87/cognitive-battery/issues) in this repo, or get live help on [Gitter](https://gitter.im/cognitive-battery/Lobby).

## TODO

In no particular order...

**General**
- Create an analysis tool for summarizing/aggregating participant data for
the different battery tasks. The goal is to output summary data that is ready
for statistics
- Compile an executable binary version of the entire battery
    - This will make it possible run the battery without dealing with Python
     module installation
- Cross-platform support (OSX and Debian based GNU/Linux)
- **Major** code cleanup across the board

**Tasks**
- Improve support for different screen resolutions
- Streamline the process of adding new tasks
    - Automatically import all task modules
    - Automatically handle task running/saving
    - No longer need to edit the main battery file to add your own tasks
- Save subject data to non-proprietary file format instead of Excel
    - Likely to be CSV, but open to suggestions
- Add more tasks...

## Changelog

You can view the change log/release notes [here](CHANGELOG.md)


CHANGELOG.md
# Change Log

## [3.2.1](https://github.com/sho-87/cognitive-battery/releases/tag/3.2.1) *(2018-11-15)*

**General**
- Code formatting.
- Added Update UI that checks GitHub for new releases.

**Tasks**
- Added analysis for Eriksen Flanker task.

**Bug Fixes**
- Fixed a bug where leading zeros in subject numbers are missing in aggregated data.

## [3.2.0](https://github.com/sho-87/cognitive-battery/releases/tag/3.2.0) *(2017-09-14)*

**General**
- Added tooltips to the Settings window items to clarify the purpose of each option.
- App-wide values moved to separate module (e.g. URLs).
- Streamlined the setting of default values for task options.
- Added placeholder scripts for aggregating task data.
- Fixed a few typos.

**Tasks**
- Added Eriksen Flanker task.

**Bug Fixes**
- Fixed a bug where settings for new tasks are not read/saved if `settings.ini` already exists.

## [3.1.0](https://github.com/sho-87/cognitive-battery/releases/tag/3.1.0) *(2017-07-06)*

**General**
- Added a menu item to check for latest releases of the battery (redirects to GitHub releases)

**Tasks**
- Added information about each task to [tasks/README.md](tasks/README.md).
- Added options to adjust Raven's trial count and starting position to Settings menu.
- Added option to play an audio beep after each task (default: True)

**Bug Fixes**
- Fixed a bug where the Raven's task was loading the images in an incorrect order.

## [3.0.0](https://github.com/sho-87/cognitive-battery/releases/tag/3.0.0) *(2017-07-03)*

**General**
- Migrated the code base over to Python 3.x. Version 3+ of the battery will be based on Python 3. Python 2 users can still use v2 of the battery, however, it will no longer be maintained.

**Tasks**
- Added option to set the number of blocks for the Sternberg task in the Settings menu.

**Bug Fixes**
- Fixed a bug where duplicate filenames were being checked for (instead of duplicate subject numbers) when starting new sessions.

## [2.0.0](https://github.com/sho-87/cognitive-battery/releases/tag/2.0.0) *(2017-07-03)*

**General**
- Updated `PyQt` requirement to version 5. This is a breaking change as PyQt5.x will be required for current and future versions of the battery. PyQt4 is still supported in version 1.x of the battery.
- The main battery window class has now been moved into its own module.

**User Interface**
- Added a project manager. Data and settings are saved on an individual project basis to their own separate directories.
- Added Settings window sections for manually specifying options for specific tasks. Currently, an option has been added to adjust the block count of the Attention Network Test, but the framework is in place to allow task options to be set by the user for all tasks.

**Bug Fixes**
- Fixed `libpng` warning with task images. The warning should no longer appear with the updated images in the battery.

## [1.1.0](https://github.com/sho-87/cognitive-battery/releases/tag/1.1.0) *(2016-10-09)*

**General**
- Moved a number of text, image and background display functions into their own package (`utils`)

**User Interface**
- Created separate class for UI layout and import into main script
- Changed UI base class to `QTMainWindow`
- Created conversion script for building a Python module from UI file
- Added Menu and Status bar to UI
- Added application icons
- Added global Settings window
- Application now stores (and restores) the size and position of windows

**Tasks**
- All tasks (and their images) moved to separate directory
- Improved integration with pygame
- SART timing has been fine tuned to better match the original publication
- Added the Sternberg Task
- Short version of the ANT has been removed

**Bug Fixes**
- Up/Down buttons now correctly disable if random order is selected
- Pygame window now closes if the main battery UI is closed
- CPU time and memory were being used up needlessly in certain tasks. This has been improved
- Error now correctly displays if no tasks have been chosen
- Pygame windows and backgrounds are now passed around the different tasks correctly
- User input is no longer (pre)registered in digit span during digit display

## [1.0.0](https://github.com/sho-87/cognitive-battery/releases/tag/1.0) *(2015-11-21)*
- Initial release


tasks/ravens.py
import time
import pandas as pd
import numpy as np
import pygame

from pygame.locals import *
from os import listdir
from os.path import join, dirname, realpath, splitext
from sys import exit


class Ravens(object):
    def __init__(self, screen, background, start=13, numTrials=12):
        # Get the pygame display window
        self.screen = screen
        self.background = background

        # sets font and font size
        self.instructionsFont = pygame.font.SysFont("arial", 20)

        # get screen info
        self.screen_x = self.screen.get_width()
        self.screen_y = self.screen.get_height()

        # Fill background
        self.background.fill((255, 255, 255))
        pygame.display.set_caption("Ravens Progressive Matrices")
        pygame.mouse.set_visible(0)

        # set number of trials
        self.numTrials = numTrials
        self.stimDuration = 60000
        self.ITI = 1000

        # get images
        self.directory = dirname(realpath(__file__))
        self.imagePath = self.directory + "\images\\Ravens\\"

        # store all filenames in the images path to a list
        self.dirImages = [
            f
            for f in listdir(self.imagePath)
            if splitext(join(self.imagePath, f))[-1] == ".png"
        ]

        self.dirImages = sorted(self.dirImages)

        # only load the desired number/set of images
        self.images = []
        for i in range(start - 1, start + numTrials - 1):
            self.images.append(pygame.image.load(self.imagePath + self.dirImages[i]))

        # get image size
        self.stimH = self.images[0].get_rect().height
        self.stimW = self.images[0].get_rect().width

        # load practice image
        self.practiceImage = pygame.image.load(
            self.imagePath + "practice\\practice.png"
        )

        # load instructions page example images
        self.img_example = pygame.image.load(self.imagePath + "practice\\example.png")
        self.exampleW = self.img_example.get_rect().width

        self.img_example_answers = pygame.image.load(
            self.imagePath + "practice\\example_answers.png"
        )
        self.exampleAnswersW = self.img_example_answers.get_rect().width

        # ravens set 2 answers
        self.correctAnswers = np.array(
            [
                5,  # 1
                1,  # 2
                7,  # 3
                4,  # 4
                3,  # 5
                1,  # 6
                6,  # 7
                1,  # 8
                8,  # 9
                4,  # 10
                5,  # 11
                6,  # 12
                2,  # 13
                1,  # 14
                2,  # 15
                4,  # 16
                6,  # 17
                7,  # 18
                3,  # 19
                8,  # 20
                8,  # 21
                7,  # 22
                6,  # 23
                3,  # 24
                7,  # 25
                2,  # 26
                7,  # 27
                5,  # 28
                6,  # 29
                5,  # 30
                4,  # 31
                8,  # 32
                5,  # 33
                1,  # 34
                3,  # 35
                2,  # 36
            ]
        )

        # create output dataframe
        self.allData = pd.DataFrame()
        self.trial = np.arange(1, numTrials + 1)
        self.allData["trial"] = self.trial
        self.stimNum = np.arange(start, start + numTrials)
        self.allData["image"] = self.stimNum
        self.answerSubset = self.correctAnswers[start - 1 : start + numTrials - 1]
        self.allData["correctAnswer"] = self.answerSubset

    def pressSpace(self, x, y):
        self.space = self.instructionsFont.render(
            "(Press spacebar when ready)", 1, (0, 0, 0)
        )
        self.screen.blit(self.space, (x, y))

    def displayTrial(self, i, data, type):
        # clear the event queue before checking for responses
        pygame.event.clear()

        if type == "main":
            self.curImage = self.images[i]
        elif type == "practice":
            self.curImage = self.practiceImage

        self.baseTime = int(round(time.time() * 1000))
        while int(round(time.time() * 1000)) - self.baseTime < self.stimDuration:
            self.endTime = int(round(time.time() * 1000))

            data.set_value(i, "userAnswer", "NA")
            data.set_value(i, "RT", "NA")

            for event in pygame.event.get():
                if event.type == KEYDOWN and event.key == K_F12:
                    pygame.quit()
                    exit()
                elif event.type == KEYDOWN:
                    if event.key == K_1:
                        data.set_value(i, "userAnswer", "1")
                        data.set_value(
                            i,
                            "RT",
                            str((float(self.endTime) - float(self.baseTime)) / 1000),
                        )
                        return 0
                    elif event.key == K_2:
                        data.set_value(i, "userAnswer", "2")
                        data.set_value(
                            i,
                            "RT",
                            str((float(self.endTime) - float(self.baseTime)) / 1000),
                        )
                        return 0
                    elif event.key == K_3:
                        data.set_value(i, "userAnswer", "3")
                        data.set_value(
                            i,
                            "RT",
                            str((float(self.endTime) - float(self.baseTime)) / 1000),
                        )
                        return 0
                    elif event.key == K_4:
                        data.set_value(i, "userAnswer", "4")
                        data.set_value(
                            i,
                            "RT",
                            str((float(self.endTime) - float(self.baseTime)) / 1000),
                        )
                        return 0
                    elif event.key == K_5:
                        data.set_value(i, "userAnswer", "5")
                        data.set_value(
                            i,
                            "RT",
                            str((float(self.endTime) - float(self.baseTime)) / 1000),
                        )
                        return 0
                    elif event.key == K_6:
                        data.set_value(i, "userAnswer", "6")
                        data.set_value(
                            i,
                            "RT",
                            str((float(self.endTime) - float(self.baseTime)) / 1000),
                        )
                        return 0
                    elif event.key == K_7:
                        data.set_value(i, "userAnswer", "7")
                        data.set_value(
                            i,
                            "RT",
                            str((float(self.endTime) - float(self.baseTime)) / 1000),
                        )
                        return 0
                    elif event.key == K_8:
                        data.set_value(i, "userAnswer", "8")
                        data.set_value(
                            i,
                            "RT",
                            str((float(self.endTime) - float(self.baseTime)) / 1000),
                        )
                        return 0

            self.screen.blit(self.background, (0, 0))
            self.screen.blit(
                self.curImage,
                (
                    self.screen_x / 2 - self.stimW / 2,
                    self.screen_y / 2 - self.stimH / 2,
                ),
            )

            self.timeLeft = (
                self.stimDuration / 1000 - (self.endTime - self.baseTime) / 1000
            )
            # convert seconds to time format
            self.timer = time.strftime("%M:%S", time.gmtime(self.timeLeft))

            self.timerText = self.instructionsFont.render(
                "Time left: " + str(self.timer), 1, (0, 0, 0)
            )
            self.timerW = self.timerText.get_rect().width
            self.screen.blit(
                self.timerText,
                (self.screen_x / 2 - self.timerW / 2, self.screen_y / 2 + 400),
            )

            pygame.display.flip()

    def run(self):
        # Instructions
        self.screen.blit(self.background, (0, 0))

        self.title = self.instructionsFont.render(
            "Raven's Progressive Matrices", 1, (0, 0, 0)
        )
        self.titleW = self.title.get_rect().width
        self.screen.blit(
            self.title, (self.screen_x / 2 - self.titleW / 2, self.screen_y / 2 - 400)
        )

        self.line1 = self.instructionsFont.render(
            "You will see a grid of items with one item missing:", 1, (0, 0, 0)
        )
        self.screen.blit(self.line1, (100, self.screen_y / 2 - 350))

        self.screen.blit(
            self.img_example,
            (self.screen_x / 2 - self.exampleW / 2, self.screen_y / 2 - 300),
        )

        self.line2 = self.instructionsFont.render(
            "There will be a set of 8 possible answer options:", 1, (0, 0, 0)
        )
        self.screen.blit(self.line2, (100, self.screen_y / 2 - 100))

        self.screen.blit(
            self.img_example_answers,
            (self.screen_x / 2 - self.exampleAnswersW / 2, self.screen_y / 2 - 50),
        )

        self.line3 = self.instructionsFont.render(
            "Determine which option is the missing item.", 1, (0, 0, 0)
        )
        self.screen.blit(self.line3, (100, self.screen_y / 2 + 150))

        self.line4 = self.instructionsFont.render(
            "In example above, the correct answer is 4.", 1, (0, 0, 0)
        )
        self.screen.blit(self.line4, (100, self.screen_y / 2 + 180))

        self.line5 = self.instructionsFont.render(
            "Select your answer by pressing the corresponding number on the keyboard.",
            1,
            (0, 0, 0),
        )
        self.screen.blit(self.line5, (100, self.screen_y / 2 + 250))

        self.line6 = self.instructionsFont.render(
            "You will have 1 minute to complete each question.", 1, (0, 0, 0)
        )
        self.screen.blit(self.line6, (100, self.screen_y / 2 + 280))

        self.pressSpace(100, (self.screen_y / 2) + 350)

        self.instructions = True
        while self.instructions:
            for event in pygame.event.get():
                if event.type == KEYDOWN and event.key == K_SPACE:
                    self.instructions = False
                elif event.type == KEYDOWN and event.key == K_F4:
                    return pd.DataFrame()
                elif event.type == KEYDOWN and event.key == K_F12:
                    pygame.quit()
                    exit()

            pygame.display.flip()

        # Instructions Practice
        self.instructionsPractice = True
        while self.instructionsPractice:
            for event in pygame.event.get():
                if event.type == KEYDOWN and event.key == K_SPACE:
                    self.instructionsPractice = False
                elif event.type == KEYDOWN and event.key == K_F4:
                    return pd.DataFrame()
                elif event.type == KEYDOWN and event.key == K_F12:
                    pygame.quit()
                    exit()

                self.screen.blit(self.background, (0, 0))
                self.practiceInstructions = self.instructionsFont.render(
                    "We will begin with a practice trial...", 1, (0, 0, 0)
                )
                self.screen.blit(self.practiceInstructions, (100, self.screen_y / 2))

                self.pressSpace(100, (self.screen_y / 2) + 100)

                pygame.display.flip()

        # Practice trials
        self.practiceData = pd.DataFrame()
        self.displayTrial(0, self.practiceData, "practice")

        # Practice feedback screen
        self.screen.blit(self.background, (0, 0))

        if self.practiceData.at[0, "userAnswer"] == "2":
            self.feedbackLine = self.instructionsFont.render("Correct", 1, (0, 255, 0))
        else:
            self.feedbackLine = self.instructionsFont.render(
                "Incorrect", 1, (255, 0, 0)
            )

        self.feedbackLineH = self.feedbackLine.get_rect().height
        self.feedbackLineW = self.feedbackLine.get_rect().width

        self.screen.blit(
            self.feedbackLine,
            (
                self.screen_x / 2 - self.feedbackLineW / 2,
                self.screen_y / 2 - self.feedbackLineH / 2,
            ),
        )

        pygame.display.flip()

        # show feedback screen for 2 seconds
        self.baseTime = int(round(time.time() * 1000))
        while int(round(time.time() * 1000)) - self.baseTime < 2000:
            pass

        # Instructions Practice End
        self.practiceEndScreen = True
        while self.practiceEndScreen:
            for event in pygame.event.get():
                if event.type == KEYDOWN and event.key == K_SPACE:
                    self.practiceEndScreen = False

            self.screen.blit(self.background, (0, 0))
            self.practiceEndLine = self.instructionsFont.render(
                "We will now begin the main trials...", 1, (0, 0, 0)
            )
            self.screen.blit(self.practiceEndLine, (100, self.screen_y / 2))

            self.pressSpace(100, (self.screen_y / 2) + 100)

            pygame.display.flip()

        # Main task
        for i in range(self.numTrials):
            self.displayTrial(i, self.allData, "main")

            if self.allData.at[i, "userAnswer"] == str(
                self.allData.at[i, "correctAnswer"]
            ):
                self.allData.set_value(i, "correct", 1)
            else:
                self.allData.set_value(i, "correct", 0)

            self.baseTime = int(round(time.time() * 1000))
            while int(round(time.time() * 1000)) - self.baseTime < self.ITI:
                self.screen.blit(self.background, (0, 0))
                pygame.display.flip()

        # rearrange dataframe
        self.columns = [
            "trial",
            "image",
            "correctAnswer",
            "userAnswer",
            "correct",
            "RT",
        ]
        self.allData = self.allData[self.columns]

        # End screen
        self.endScreen = True
        while self.endScreen:
            for event in pygame.event.get():
                if event.type == KEYDOWN and event.key == K_SPACE:
                    self.endScreen = False

            self.screen.blit(self.background, (0, 0))
            self.endLine = self.instructionsFont.render("End of task.", 1, (0, 0, 0))
            self.screen.blit(self.endLine, (100, self.screen_y / 2))

            self.pressSpace(100, (self.screen_y / 2) + 100)

            pygame.display.flip()

        print("- Raven's Progressive Matrices complete")

        return self.allData
